### ISSUE TITLE
AO and Automated Mixed Precision

### ISSUE THREAD

**Issue by bhack** [pytorch/ao#1390]
Can we clarify in the readme what are the best practices to use ao at inference with a pytorch AMP trainer model/checkpoint?

**Comment by vkuzo** [comment]
@bhack , would love to learn some more context - is there an issue when you use the checkpoint directly?

**Comment by bhack** [comment]
Yes exactly it is the case when we use a checkpoint trained with AMP and we are entering in the AO world to optimize it for inference.

**Comment by bhack** [comment]
Generally in the repo before AO I just see `model.eval().to(torch.bflaot16)`. 
Is this the best practice? 
If yes we could add some notes in the documentation.

**Comment by vkuzo** [comment]
Sorry for the late response, I was on holiday last week.  There are no special things you need to do when using amp + torchao for converting the model to inference - just use torchao according to the documentation.  Would you mind clarifying your question if that does not answer it?

**Comment by bhack** [comment]
Just to make an example from the Doc is not clear if this is the best practice or not:
```python
       model.eval()
       quantize_(model, int8_dynamic_activation_int8_weight())
       unwrap_tensor_subclass(model)
       autoquant(model, qtensor_class_list=DEFAULT_FLOAT_AUTOQUANT_CLASS_LIST,min_sqnr=40)
       with torch.amp.autocast(enabled=use_amp, dtype=input_dtype, device_type="cuda"):
           ......
           torch.export.export(....)
           aoti_compile_and_package(...)
```

**Comment by vkuzo** [comment]
I see, looks like your question is whether using autocast for quantized model inference is supported / recommended.  I'll tag @jerryzh168  on this one.

**Comment by jerryzh168** [comment]
> Just to make an example from the Doc it is not clear if this is the best practice or not:
> 
> ```python
>        model.eval()
>        quantize_(model, int8_dynamic_activation_int8_weight())
>        unwrap_tensor_subclass(model)
>        with torch.amp.autocast(enabled=use_amp, dtype=input_dtype, device_type="cuda"):
>            ......
>            torch.export.export(....)
>            aoti_compile_and_package(...)
> ```

we haven't tested this, but this seems reasonable to me, but
1. the quantized kernel should be able to handle the activation dtype from autocast
2. for inference, people typically already starts with these lower precision dtypes like bfloat16, so not sure what is the use case here

**Comment by jerryzh168** [comment]
> For 2: Do you mean `model.eval().to(torch.bfloat16)` without AMP context?

Yes

**Comment by bhack** [comment]
Ok, it think it is better to add a note somewhere as it is quite common to have models trained with AMP.

**Comment by jerryzh168** [comment]
> Ok, it think it is better to add a note somewhere as it is quite common to have models trained with AMP.

Do you know the e2e flow for a model trained with AMP? do people
1. train with AMP, save the state_dict
2. load the state_dict and quantize the model etc.?

**Comment by bhack** [comment]
I think users want to know if they need to use the amp context on a quantized model or not as the documentation still suggests to use it for inference in the amp cases:
https://pytorch.org/docs/stable/amp.html

**Comment by jerryzh168** [comment]
> I think users want to know if they need to use the amp context on a quantized model or not as the documentation still suggests to use it for inference in the amp cases: [pytorch.org/docs/stable/amp.html](https://pytorch.org/docs/stable/amp.html)

I see, I think we could add a note to say that amp is typically not required in the torchao context anymore, since we typically start with a bfloat16 or float16 model

But if people want to run inference for a model trained with AMP, I'd suggest to start with saving the trained model and load the model: https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html#saving-resuming and quantize. but we'll need to test this path before updating the README

**Comment by bhack** [comment]
Without using `model.eval().to(torch.bfloat16)` and removing amp (also if trained with amp) autoquant + export is running.

**Comment by bhack** [comment]
Starting `float32` with `float32` input after a lot of autotuning is failing

```python
  File "/workspace/model/feature_extractor.py", line 323, in forward
    x = blk(x)
        ^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1749, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1760, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/model/feature_extractor.py", line 295, in forward
    x = input * self.skip_scale + self.self_attention(x)
                                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1749, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1760, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/model/feature_extractor.py", line 252, in forward
    x = self.act(self.conv2d(x))
                 ^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1749, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1760, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/conv.py", line 554, in forward
    return self._conv_forward(input, self.weight, self.bias)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/conv.py", line 549, in _conv_forward
    return F.conv2d(
           ^^^^^^^^^
RuntimeError: Input type (c10::Half) and bias type (float) should be the same
```

**Comment by jerryzh168** [comment]
> Starting `float32` with `float32` input after a lot of autotuning is failing
> 
>   File "/workspace/model/feature_extractor.py", line 323, in forward
>     x = blk(x)
>         ^^^^^^
>   File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1749, in _wrapped_call_impl
>     return self._call_impl(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1760, in _call_impl
>     return forward_call(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspace/model/feature_extractor.py", line 295, in forward
>     x = input * self.skip_scale + self.self_attention(x)
>                                   ^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1749, in _wrapped_call_impl
>     return self._call_impl(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1760, in _call_impl
>     return forward_call(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspace/model/feature_extractor.py", line 252, in forward
>     x = self.act(self.conv2d(x))
>                  ^^^^^^^^^^^^^^
>   File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1749, in _wrapped_call_impl
>     return self._call_impl(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1760, in _call_impl
>     return forward_call(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/conv.py", line 554, in forward
>     return self._conv_forward(input, self.weight, self.bias)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/conda/lib/python3.11/site-packages/torch/nn/modules/conv.py", line 549, in _conv_forward
>     return F.conv2d(
>            ^^^^^^^^^
> RuntimeError: Input type (c10::Half) and bias type (float) should be the same

@bhack can you post the repro for autoquant here

**Comment by bhack** [comment]
I am not on that model anymore so I don't think I can recover that env quickly. 

But an official doc entry in a markdown file  on how to handle amp on AO it will still be useful.
